### ISSUE TITLE
With magento (1) in htdocs subdirectory,  creation of local.xml failed or in wrong location

### ISSUE THREAD

**Issue by kkrieger85** [drud/ddev#2291]
**Describe the bug**
Creation of `local.xml` failed. selected docroot not added to path

Example error message: `Unable to create settings file '/home/kkrieger/htdocs/PROJECT/app/etc/local.xml': open /home/kkrieger/htdocs/PROJECT/app/etc/local.xml: no such file or directory 
`

BUT: I can run `ddev start` and the project is running 
**Only issue is failing creation of the config file**

**To Reproduce**
1. Go to your project folder 
2. `ddev config`
3. Accept projectname 
4. existing `htdocs` is recognized, accept `htdocs` as docroot 
5. select `magento` from Project Type list
=> creation of local.xml failed
6. you should see that ddev tried to create the local.xml in NOT htdocs docroot in which the magento code is located

**Expected behavior**
Config file should be saved to PROJECT_PATH + previously selected docroot + `app/etc/local.xml`

**Version and configuration information (please complete the following information):**
 - Ubuntu 20.04 LTS, 64 bit

 docker version                                                                                                                                                                                      ✔  6705  09:46:24
Client:
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.13.8
 Git commit:        afacb8b7f0
 Built:             Wed Mar 11 23:42:35 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.8
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 22:48:33 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.3-0ubuntu2
  GitCommit:        
 runc:
  Version:          spec: 1.0.1-dev
  GitCommit:        
 docker-init:
  Version:          0.18.0
  GitCommit: 

DDEV-Local version      v1.13.0                                
commit                  v1.13.0                                
db                      drud/ddev-dbserver-mariadb-10.2:v1.13.0
dba                     phpmyadmin/phpmyadmin:5                
ddev-ssh-agent          drud/ddev-ssh-agent:v1.13.0            
docker                  19.03.8                                
docker-compose          1.25.0-rc4                             
os                      linux                                  
router                  drud/ddev-router:v1.13.0               
web                     drud/ddev-webserver:v1.13.0  

config.yaml

APIVersion: v1.13.0
name: werbekoenig
type: magento
docroot: htdocs
php_version: "5.6"
webserver_type: nginx-fpm
router_http_port: "80"
router_https_port: "443"
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
nfs_mount_enabled: false
provider: default
use_dns_when_possible: true
timezone: ""

 - Do you have any custom configuration (nginx, php, mysql) in the .ddev folder? 
no

**Additional context**
project structure:
/home/kkrieger/htdocs/PROJECT/
|- .ddev
|- composer.json
|- htdocs
  | - index.php
|- vendor

**Comment by rfay** [comment]
On mine it creates the local.xml just fine... in the wrong place. It creates it in <projectroot>/app/etc/local.xml instead of <projectroot>/htdocs/app/etc/local.xml, where I assume it should be created?

I note that it also doesn't autodetect the project type when magento is in a docroot, which is probably related to this. 

I have no idea why it can't create the file in your environment, permissions? HOWEVER... you say 

> Example error message: Unable to create settings file '/home/kkrieger/htdocs/PROJECT/app/etc/local.xml': open /home/kkrieger/htdocs/PROJECT/app/etc/local.xml: no such file or directory

And it's trying to put it in the wrong place right?

**Comment by kkrieger85** [comment]
I checked it again with another project:

It created `app/etc/.gitignore` but not in selected webroot

Message: `No /home/kkrieger/htdocs/PROJECT/app/etc/local.xml file exists, creating one 
Unable to create settings file '/home/kkrieger/htdocs/PROJECT/app/etc/local.xml': open /home/kkrieger/htdocs/PROJECT/app/etc/local.xml: no such file or directory `

`/home/kkrieger/htdocs/PROJECT/app/etc/local.xml: no such file or directory` << is right, the path should be `/home/kkrieger/htdocs/PROJECT/htdocs/app/etc/local.xml: no such file or directory`

Also autodetection did not work at all. ddev config proposes "wordpress" as detected code


May it be that the issue is related to my personal projects path `/home/kkrieger/htdocs/PROJECT` which already includes `htdocs`

**Comment by rfay** [comment]
I've never seen a use of htdocs like that, but don't think that's likely the problem. I think it's more likely that the autodetect of magento-in-subdir is just failing.

**Comment by rfay** [comment]
In ddev v1.16+ there is a bit of improved docroot detection, but I don't think it would affect this one.

**Comment by Flyingmana** [comment]
I had just the same issue with the `local.xml` not beeing created and using the wrong path.
I think this are separate Issues.

For the file not beeing created, on the second run of `ddev config` it got created, so I think it is related to the directory not existing initially.
I think the code has an edgecase to only create the directory and not the file, if the directory does not exist yet.
This could either be a logic error, or some weird timing issue.

My environment:

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.1 LTS
Release:        20.04
Codename:       focal

$ uname -a
Linux DESKTOP-JU5UTRU 4.19.128-microsoft-standard #1 SMP Tue Jun 23 12:58:10 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

```

I run WSL2 on Windows 10 Pro, Version 2004

**Comment by rfay** [comment]
@Flyingmana you think your issue is not the same as this one? In that case can you please open a new issue and explain yours very clearly?

**Comment by Flyingmana** [comment]
Iam pretty sure my Issue is the same as the initial reported one.
Maybe @kkrieger85 can validate " **failing creation of the config file**" only happens, if the directory does not exist before.

**Comment by rfay** [comment]
So @Flyingmana you're also working with magento 1?

**Comment by Flyingmana** [comment]
Yes

**Comment by kkrieger85** [comment]
ACK, will check tomorrow

**Comment by kkrieger85** [comment]
Generation works fine in ddev 1.16.5.

My Steps:
```
ddev --version
ddev version v1.16.5

git clone OpenMage
cd OpenMage

ddev config
-> yes to all 
(magento was identified correctly)
``` 

Second try:
Using an extra `htdocs` folder and composer.

Results: It seems, that detection of magento fails if its located in a custom docroot, e.g. `htdocs`. Without correct detection of magento, the app/etc/config.xml will be generated in wrong directory.

Here's my sample composer setup:

~/ddev-config-test/composer.json 
``` 
{
    "name": "kkrieger/ddev-config-test",
    "description": "only for test purpose",
    "authors": [
        {
            "name": "kkrieger (Kevin Krieger)",
            "email": "kk@kkrieger.de"
        }
    ],
     "require": {
        "aydin-hassan/magento-core-composer-installer": "*",
        "openmage/magento-lts": "*"
    },
    "extra": {
        "magento-core-package-type": "magento-source",
        "magento-root-dir": "htdocs"
    }
}
``` 

``` 
composer install 
=> openmage will be installed to htdocs folder
ddev config
=> doesn't detect magento/openmage

**Comment by rfay** [comment]
I'd love it if you could test with current ddev v1.17.1, as htdocs is now a "found" docroot name.

**Comment by kkrieger85** [comment]
Unfortunately, it doesn't work:

Preparation:
``` 
mkdir ddev-config-test
vi composer.json (with example above)
composer install
``` 


``` 
ddev config
Creating a new ddev project config in the current directory (path/to/ddev-config-test) 
Once completed, your configuration will be written to path/to/ddev-config-test/.ddev/config.yaml
 
Project name (ddev-config-test): 

The docroot is the directory from which your site is served.
This is a relative path from your project root at path/to/ddev-config-test 
You may leave this value blank if your site files are in the project root 
Docroot Location (htdocs): 
Found a php codebase at path/to/ddev-config-test/htdocs. 
Project Type [backdrop, drupal6, drupal7, drupal8, drupal9, laravel, magento, magento2, php, shopware6, typo3, wordpress] (php): magento
No path/to/ddev-config-test/app/etc/local.xml file exists, creating one 
Unable to create settings file 'path/to/ddev-config-test/app/etc/local.xml': open path/to/ddev-config-test/app/etc/local.xml: no such file or directory 
``` 

So ddev config seems to ignore previousely set docroot location

**Comment by rfay** [comment]
It's not ignoring it, it reports that it used it, "Found a php codebase at path/to/ddev-config-test/htdocs", but apparently it's not looking in there. 

Thanks for the re-check on this, it will make it easier to get fixed up. Thanks!

**Comment by sreichel** [comment]
Hi,

run into same issue. Shouldn\`t `settingsFileBasePath`  include `docroot` form config.yaml or from composer.json `magento-root-dir`?

```
// setMagentoSiteSettingsPaths sets the paths to settings.php for templating.
func setMagentoSiteSettingsPaths(app *DdevApp) {
	settingsFileBasePath := app.AppRoot
	app.SiteSettingsPath = filepath.Join(settingsFileBasePath, "app", "etc", "local.xml")
}
```

**Comment by rfay** [comment]
Thanks @sreichel - do you want to do a PR to fix it? Easy to test on gitpod and easy to test locally. This definitely deserves to be fixed in this release cycle. Really before now :) https://github.com/drud/ddev/blob/master/docs/developers/building-contributing.md

**Comment by rfay** [comment]
Fixed in #3481 - this wasn't just a magento 1 problem, also magento2
